lib/kmer_hash.hh

```diff
@@ -79,6 +79,217 @@ HashIntoType _hash_murmur(const std::string& kmer);
 HashIntoType _hash_murmur(const std::string& kmer,
                           HashIntoType& h, HashIntoType& r);
 HashIntoType _hash_murmur_forward(const std::string& kmer);
+
+/**
+ * \class Kmer
+ *
+ * \brief Hold the hash values corresponding to a single k-mer.
+ *
+ * This class stores the forward, reverse complement, and
+ * uniqified hash values for a given k-mer. It also defines
+ * some basic operators and a utility function for getting
+ * the string representation of the sequence. This is meant
+ * to replace the original inelegant macros used for hashing.
+ *
+ * \author Camille Scott
+ *
+ * Contact: camille.scott.w@gmail.com
+ *
+ */
+class Kmer
+{
+
+public:
+
+    /// The forward hash
+    HashIntoType kmer_f;
+    /// The reverse (complement) hash
+    HashIntoType kmer_r;
+    /// The uniqified hash
+    HashIntoType kmer_u;
+
+    /** @param[in]   f forward hash.
+     *  @param[in]   r reverse (complement) hash.
+     *  @param[in]   u uniqified hash.
+     */
+    Kmer(HashIntoType f, HashIntoType r, HashIntoType u)
+    {
+        kmer_f = f;
+        kmer_r = r;
+        kmer_u = u;
+    }
+
+    /// @warning The default constructor builds an invalid k-mer.
+	Kmer()
+	{
+		kmer_f = kmer_r = kmer_u = 0;
+	}
+
+    /// Allows complete backwards compatibility
+    operator HashIntoType() const { return kmer_u; }
+
+    bool operator< (const Kmer &other) const {
+        return kmer_u < other.kmer_u;
+    }
+
+    std::string get_string_rep(WordLength K) {
+        return _revhash(kmer_u, K);
+    }
+
+};
+
+/**
+ * \Class KmerFactory
+ *
+ * \brief Build complete Kmer objects.
+ *
+ * The KmerFactory is a simple construct to emit complete
+ * Kmer objects. The design decision leading to this class
+ * stems from the issue of overloading the Kmer constructor
+ * while also giving it a K size: you get ambiguous signatures
+ * between the (kmer_u, K) and (kmer_f, kmer_r) cases. This
+ * implementation also allows a logical architecture wherein
+ * KmerIterator and Hashtable inherit directly from KmerFactory,
+ * extending the latter's purpose of "create k-mers" to 
+ * "emit k-mers from a sequence" and "create and store k-mers".
+ *
+ * \author Camille Scott
+ *
+ * Contact: camille.scott.w@gmail.com
+ *
+ */
+class KmerFactory
+{
+protected:
+    WordLength _ksize;
+
+public:
+
+    KmerFactory(WordLength K): _ksize(K) {}
```

Comment:
Whoops, found another cppcheck warning:
`Class 'KmerFactory' has a constructor with 1 argument that is not explicit.`